Added e2e tests for `page.tag.attached` webhook

leonyangela opened the pull request:
ref #15537

Got some code for us? Awesome 🎊!

Please include a description of your change & check your PR against this list, thanks!

- [x] There's a clear use-case for this code change, explained below
- [x] Commit message has a short title & references relevant issues
- [ ] The build will pass (run `yarn test:all` and `yarn lint`)

We appreciate your contribution!

Also, if you'd be interested in writing code like this for us more regularly, we're hiring:
https://careers.ghost.org

naz commented on ghost/core/test/e2e-webhooks/pages.test.js:
```diff
@@ -31,7 +43,7 @@ const buildAuthorSnapshot = (roles = false) => {
 
     if (roles) {
-        return {...authorSnapshot, roles: Array(1).fill(roleSnapshot)};
+        return { ...authorSnapshot, roles: Array(1).fill(roleSnapshot) };
```
No need for extra spaces here 🙃

leonyangela replied:
I apologize, my bad. I've resolved this, thank you.

naz commented on ghost/core/test/e2e-webhooks/pages.test.js:
```diff
@@ -56,13 +68,47 @@ const buildPageSnapshotWithTiers = ({
     };
 };
 
+const buildPageSnapshotWithTiersAndTags = ({
```
Would it make sense to reuse the `buildPageSnapshotWithTiers` method here?

naz commented on ghost/core/test/e2e-webhooks/pages.test.js:
```diff
@@ -23,6 +24,17 @@ const tierSnapshot = {
 
 const roleSnapshot = tierSnapshot;
 
+const tagSnapshot = {
```
The tag's name, slug, visibility etc. are not changing between test runs? If they are not, would be best to have snapshot capture them as is - no need to include non-dynamic properties in the snapshot matcher object 👍

naz commented on ghost/core/test/e2e-webhooks/pages.test.js:
```diff
@@ -171,4 +217,61 @@ describe('page.* events', function () {
+            })
+            .expectStatus(201);
+
+            console.log(res);
```
Probably a leftover debugging code 😄

leonyangela replied:
I apologize, my bad.

naz reviewed (changes requested):
Hey @leonyangela! Thanks for taking time to make the contribution. It's in pretty good shape just needs to polish up a few rough edges. I've left some guiding comments in the PR for the improvements. Feel free to reach out if you need more guidance 😊

leonyangela commented:
Hi @naz, Thanks for your feedback. I've made another changes. I've tried to run `yarn lint` but there is an error saying that:
```
Oops! Something went wrong! :(

ESLint: 8.25.0

No files matching the pattern "'core/server/**/*.js'" were found.
Please check for typing mistakes in the pattern.

error Command failed with exit code 2.
```
Do you have any idea for this? I also made a changes on tagSnapshot, please let me know if there's more feedback. Thank you.

leonyangela committed:
Changes on e2e tests for page.tag.attached webhook

ref #15537

🎨 Update tagSnapshot
🎨 Update buildPreviousPageSnapshotWithTiersAndTags function

leonyangela commented:
Hi @naz, I've updated the tagSnapshot and made some changes. Please let me know if there's any more feedback, cheers!

naz commented:
Hey @leonyangela , just a heads up there are some linting errors that need to be fixed. The rest of the code is heading in a right direction 👍

leonyangela commented:
Hi @naz, the lint test flow keeps failing, I have no idea on this, any suggestions?

naz reviewed (approved):
The linting passed! 🚀  You got there eventually @leonyangela , thanks for the effort and great contribution 💪

leonyangela commented:
ah great, thank you @naz.

ErisDS commented:
Hi @leonyangela  thanks so much for this PR. It has now been merged 🎉  FYI I had to rebase it as it conflicted with changes made to add the `page.tag.detached` tests.

I'm not sure if you found this through hacktoberfest, but I've added the `hacktoberfest-accepted` label to this PR to make sure it counts. We have plenty of open issues and ongoing projects for hacktoberfest and beyond. We would love to see you around again 👋

leonyangela commented:
Hi @ErisDS, thank you. Yes, I found this from hacktoberfest, thanks for adding the `hacktoberfest-accepted` label, cheers :beers:.
